rust/src/origin.rs

```diff
@@ -6,56 +6,236 @@
  * SPDX-License-Identifier: Apache-2.0 OR MIT
  */
 
-use anyhow::{bail, Result};
-use glib::translate::*;
-use glib::GString;
+use crate::cxxrsutil::*;
+use crate::treefile::Treefile;
+use anyhow::Context;
+use anyhow::Result;
+use fn_error_context::context;
 use glib::KeyFile;
-use indoc::indoc;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::pin::Pin;
 use std::result::Result as StdResult;
 
-use std::collections::{BTreeMap, BTreeSet};
-
 const ORIGIN: &str = "origin";
-const OVERRIDE_COMMIT: &str = "override-commit";
+const RPMOSTREE: &str = "rpmostree";
+const PACKAGES: &str = "packages";
+const OVERRIDES: &str = "overrides";
+
+/// The set of keys that we parse as BTreeMap and need to ignore ordering changes.
+static UNORDERED_LIST_KEYS: phf::Set<&'static str> = phf::phf_set! {
+    "packages/local",
+    "overrides/replace-local"
+};
+
+#[context("Parsing origin")]
+fn origin_to_treefile_inner(kf: &KeyFile) -> Result<Box<Treefile>> {
+    // First thing here, we remove libostree-transient bits, as those should
+    // never carry over into new origins, and we don't actually operate on those
+    // fields.
+    ostree::Deployment::origin_remove_transient_state(kf);
+
+    let mut cfg: crate::treefile::TreeComposeConfig = Default::default();
+    let refspec_str = if let Some(r) = keyfile_get_optional_string(&kf, ORIGIN, "refspec")? {
+        Some(r)
+    } else {
+        keyfile_get_optional_string(&kf, ORIGIN, "baserefspec")?
+    };
+    cfg.derive.from = refspec_str;
```

Comment:
Should we error out here if it's `None`?

Reply:
Yes, good catch!